cpp/src/utilities/cuda_memcpy.cu

```diff
@@ -0,0 +1,105 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/cuda_memcpy.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/pinned_memory.hpp>
+
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/copy.h>
+
+namespace cudf::detail {
+
+namespace {
+
+void copy_pinned(void* dst, void const* src, std::size_t size, rmm::cuda_stream_view stream)
+{
+  if (size == 0) return;
+
+  if (size < get_kernel_pinned_copy_threshold()) {
+    thrust::copy_n(rmm::exec_policy_nosync(stream),
+                   static_cast<const char*>(src),
+                   size,
+                   static_cast<char*>(dst));
+  } else {
+    CUDF_CUDA_TRY(cudaMemcpyAsync(dst, src, size, cudaMemcpyDefault, stream));
+  }
+}
+
+void copy_pageable(void* dst, void const* src, std::size_t size, rmm::cuda_stream_view stream)
+{
+  if (size == 0) return;
+
+  CUDF_CUDA_TRY(cudaMemcpyAsync(dst, src, size, cudaMemcpyDefault, stream));
+}
+
+void copy_pinned_to_device(void* dst,
+                           void const* src,
+                           std::size_t size,
+                           rmm::cuda_stream_view stream)
+{
+  copy_pinned(dst, src, size, stream);
+}
+
+void copy_device_to_pinned(void* dst,
+                           void const* src,
+                           std::size_t size,
+                           rmm::cuda_stream_view stream)
+{
+  copy_pinned(dst, src, size, stream);
+}
+
+void copy_device_to_pageable(void* dst,
+                             void const* src,
+                             std::size_t size,
+                             rmm::cuda_stream_view stream)
+{
+  copy_pageable(dst, src, size, stream);
+}
+
+void copy_pageable_to_device(void* dst,
+                             void const* src,
+                             std::size_t size,
+                             rmm::cuda_stream_view stream)
+{
+  copy_pageable(dst, src, size, stream);
+}
+
+};  // namespace
+
+void cuda_memcpy_async(
+  void* dst, void const* src, size_t size, copy_kind kind, rmm::cuda_stream_view stream)
+{
+  if (kind == copy_kind::PINNED_TO_DEVICE) {
+    copy_pinned_to_device(dst, src, size, stream);
+  } else if (kind == copy_kind::DEVICE_TO_PINNED) {
+    copy_device_to_pinned(dst, src, size, stream);
+  } else if (kind == copy_kind::PAGEABLE_TO_DEVICE) {
+    copy_pageable_to_device(dst, src, size, stream);
+  } else if (kind == copy_kind::DEVICE_TO_PAGEABLE) {
+    copy_device_to_pageable(dst, src, size, stream);
+  }
```

Comment:
```suggestion
  switch(kind) {
  case copy_kind::PINNED_TO_DEVICE:
  case copy_kind::DEVICE_TO_PINNED:
    copy_pinned(dst, src, size, stream);
  case copy_kind::PAGEABLE_TO_DEVICE:
  case copy_kind::DEVICE_TO_PAGEABLE:
  case default:
    copy_pageable(dst, src, size, stream);
```

but better:

```
cudaPointerAttributes src_attribs;
CUDF_CUDA_TRY(cudaPointerGetAttributes(... &src_attribs));
cudaPointerAttributes dst_attribs;
CUDF_CUDA_TRY(cudaPointerGetAttributes(... &dst_attribs));
bool pageable = ((src_attribs.cudaMemoryType == cudaMemoryTypeUnregistered) or 
  (dst_attribs.cudaMemoryType == cudaMemoryTypeUnregistered));
if (pageable and size < get_kernel_pinned_copy_threshold()) {
  thrust::copy_n(rmm::exec_policy_nosync(stream),
                 static_cast<const char*>(src)
                 size,
                 static_cast<char*>(dst));
} else {
  CUDF_CUDA_TRY(cudaMemcpyAsync(dst, src, size, cudaMemcpyDefault, stream));
}
```

Reply:
I was told that `cudaPointerGetAttributes` is not trivial, so I'm trying to avoid calling it for every copy. Also, FWIW tying the strategy to the memory type prevents callers from manually overriding the strategy.
Current API _is_ awkward to use when copying from an existing `cudf::host_vector`, so I'm not sure what's the best option here.